keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -0,0 +1,276 @@
+Requests at higher priority will __always__ be executed prior to requests at lower.
+3. *Distributing*: We poll requests from each priority band's queue by
+[Weighted Round Robin](https://en.wikipedia.org/wiki/Weighted_round_robin)
+algorithm. If the request's matched bucket's weight is higher, its request
+will have more chance to be executed.
+
+Each stage works separately and solves separate goals.
+
+```
+                    +-----------+  enqueued  +--------------+  pending  +--------------+ polled by WRR
+Inbound Request --> | Reserving | -------->  | Prioritizing | --------> | Distributing |  ------->  Executed
+                    +-----------+            +--------------+           +--------------+
+              sufficient |             insufficient |                          |
+          reserved quota |        STSB shared quota |                          |
+                         v                          v                          v
+                 Executed Immediately           Rejected                    Time-out
+```
+
+
+##### API Model and APIServer flags
+
+
+
+```go
+type Bucket struct {
+    ...
+
+    // ReservedQuota defines a max concurrency of requests can run under guarantee w/o competing.
+    ReservedQuota int
+
+    // SharedQuota defines a shared concurrency added to the system's shared bandwidth.
+    SharedQuota int
+
+    // Priority assigns the bucket into a certain priority band.
+    Priority PriorityBand // Ranged in 1...N
+
+    // Weight defines how much weight its matched requests have in the queue.
+    Weight int
+}
+
+type BucketBinding struct {
+    ...
+
+    // Rules defines how we bind request to a bucket.
+    Rules []BucketBindingRule // WIP
+
+    // BucketRef references a bucket.
+    BucketRef *Bucket
+}
+```
+
+Also, we will add two optional flags to apiserver:
+
+- *--initial-extra-shared-quota*: default to 200-ish. Extra shared quota will be consumed when a request fails to get any reserved or shared quota from the defined buckets. Or we can say it's a "lowest" priority which is capable of running requests from any priority band. All requests queued in this "lowest" priority will be weighted evenly.
+- *--extra-shared-quota-auto-tuning*: forced to false in the alpha stage. If enabled, the extra bucket quota will be auto-tuned based on an in-memory moving average.
+
+
+##### Implementation
+
+See a poc implementation at: [yue9944882/inflight-rate-limit](https://github.com/yue9944882/inflight-rate-limit)
+
+This working system works as a filter handler in generic apiserver, so it
+works both for kube-apiserver and aggregated apiservers:
+- The filter list-watches *Bucket* and *BucketBinding* API models and reloads
+ the buckets dynamically
+on notification of changes.
+- The filter consists of one *quota producer* and one *queue drainer*:
+  - *quota producer*: list-watches *Bucket* API models and records status
+  of each bucket's reserved/shared quota in a map. The bucket's quota will
```

Comment:
Clearly we don't actually write this information to etcd, correct?  Conceptually it is helpful and perhaps a write every X seconds, but not on every request

Reply:
Given how dynamically this changes, writing it to etcd may be pretty misleading in my opinion.

Reply:
> Clearly we don't actually write this information to etcd, correct?

yes, the filter works in each kube-apiserver separately. so writing the real-time status into etcd could lead to extreme RV racing. but we are tracking this by prometheus metrics.
```
apiserver_bucket_used_quota{type="[reserved|shared]", bucket_name="<name>", priority_band="<Top|...|Lowest>"}
```

> Conceptually it is helpful and perhaps a write every X seconds, but not on every request

flushing them into status periodly can work. iirc AA servers also have their discovery address so  we can track bucket status for both kube-apiserver replicas and AA replicas e.g.

```
kind: Bucket
status:
  quotaUsage:
  - <[kube|AA]-apiserverDiscoveryAddress>: <count>
  ...
```